cmd/crank/beta/top/top.go

```diff
@@ -0,0 +1,252 @@
+	errWriteRow               = "cannot write row"
+	errFlushTabWriter         = "cannot flush tab writer"
+)
+
+// Cmd represents the top command.
+type Cmd struct {
+	Split bool `short:"s" name:"split" help:"Show each category separately."`
+}
+
+// Help returns help instructions for the top command.
+func (c *Cmd) Help() string {
+	return `Usage: crossplane top`
+}
+
+type topMetrics struct {
+	PodType      string
+	PodName      string
+	PodNamespace string
+	CPUUsage     int64
+	MemoryUsage  string
+}
+
+type defaultPrinterRow struct {
+	podType   string
+	namespace string
+	name      string
+	cpu       string
+	memory    string
+}
+
+func (r *defaultPrinterRow) String() string {
+	return strings.Join([]string{
+		r.podType,
+		r.namespace,
+		r.name,
+		r.cpu,
+		r.memory,
+	}, "\t") + "\t"
+}
+
+// Run runs the top command.
+func (c *Cmd) Run(k *kong.Context, logger logging.Logger) error { //nolint:gocyclo // TODO:(piotr1215) refactor to use dedicated functions
+	logger = logger.WithValues("cmd", "top")
+
+	var w = k.Stdout
+	tw := printers.GetNewTabWriter(w)
+	logger.Debug("Tabwriter header created")
+
+	// Build the config from the kubeconfig path
+	config, err := ctrl.GetConfig()
+	if err != nil {
+		return errors.Wrap(err, errKubeConfig)
+	}
+	logger.Debug("Found kubeconfig")
+
+	// Create the clientset for Kubernetes
+	k8sClientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, errCreateK8sClientset)
+	}
+	logger.Debug("Created clientset for Kubernetes")
+
+	// Create the clientset for Metrics
+	metricsClientset, err := versioned.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, errCreateMetricsClientset)
+	}
+	logger.Debug("Created clientset for Metrics")
+
+	ctx := context.Background()
+	pods, err := k8sClientset.CoreV1().Pods(metav1.NamespaceAll).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, errFetchAllPods)
+	}
+
+	crossplanePods := getCrossplanePods(pods.Items)
+	logger.Debug("Fetched all pods from all namespaces and filtered out Crossplane", "pods", crossplanePods)
+
+	for i, pod := range crossplanePods {
+		podMetrics, err := metricsClientset.MetricsV1beta1().PodMetricses(pod.PodNamespace).Get(ctx, pod.PodName, metav1.GetOptions{})
+		if err != nil {
+			return errors.Wrap(err, errAddingPodMetrics)
+		}
+		var totalCPUUsage, totalMemoryUsage int64
+		for _, container := range podMetrics.Containers {
+			totalCPUUsage += container.Usage.Cpu().ScaledValue(resource.Milli)
+			totalMemoryUsage += container.Usage.Memory().ScaledValue(resource.Mega)
+		}
+
+		crossplanePods[i].CPUUsage = totalCPUUsage
+		crossplanePods[i].MemoryUsage = fmt.Sprintf("%dMi", totalMemoryUsage)
+	}
+	logger.Debug("Metrics added to Crossplane pods")
+
+	if err != nil {
+		return errors.Wrap(err, errGetPodMetrics)
+	}
+	logger.Debug("Added metrics to Crossplane pods")
+	if c.Split {
+		printPodsByCategory(crossplanePods)
+	} else {
```

Comment:
The logic is already in the `printPodsByCategory` function. Are you suggesting moving the if block to the function as well?

Reply:
I'm referring to the content of the else block. Why break the content of the if block out into a function, but not the (similarly complex) content of the else block?

Reply:
Thank you for clarifying @negz , that makes sense now. 
I've moved the tabwriter functionality to a separate function :+1: 
This makes it a little harder to test, since function takes an implementation as parameter. What is a good _crossplane idiomatic_ way of testing such functions?

Reply:
I would use the same pattern you've used in `TestGetCrossplanePods` to test `printPodsTable` and `printPodsByCategory`. I'm not familiar with `printers.TabWriter`. Is it possible to read back what was written to it? If so, I imagine `got` in the unit tests would be whatever was written to the `TabWriter`.

Reply:
I finally settled on returning a string from the printPodsTable function and operating on in with tabwriter from the Run method. This way the function depends only on a struct and is much easier to test without needing to pass the tabwriter into it.

Reply:
Actually, the function is now refactored to take the io.writer interface. Looks like this is a bit more idiomatic way in Go than always trying to use pure functions.